File: microservice-architecture/microservice-basics/js/test/test_http.js
import http from 'node:http'

import { HttpInventoryClient } from '../src/HttpInventoryClient.js'
import { OrderServiceAsync } from '../src/OrderServiceAsync.js'

function assertEqual(expected, actual, message) {
  if (expected !== actual) {
    throw new Error(`${message} expected=${expected} actual=${actual}`)
  }
}

;(async function run() {
  let stock = 2

  const server = http.createServer((req, res) => {
    const url = new URL(req.url, 'http://127.0.0.1')
    if (url.pathname !== '/reserve') {
      res.writeHead(404)
      res.end()
      return
    }

    const sku = url.searchParams.get('sku')
    const quantity = Number(url.searchParams.get('quantity') || '0')

    if (sku === 'SKU-BOOK' && quantity > 0 && stock >= quantity) {
      stock -= quantity
      res.writeHead(200)
      res.end('OK')
      return
    }

    res.writeHead(409)
    res.end('NO_STOCK')
  })

  await new Promise((resolve) => server.listen(0, '127.0.0.1', resolve))
  const port = server.address().port

  try {
    const client = new HttpInventoryClient(`http://127.0.0.1:${port}`)
    const service = new OrderServiceAsync(client)

    const success = await service.createOrder('ORD-2001', 'SKU-BOOK', 1)
    assertEqual('CREATED', success.status, 'http status should be CREATED')

    const failed = await service.createOrder('ORD-2002', 'SKU-BOOK', 2)
    assertEqual('REJECTED', failed.status, 'http status should be REJECTED')

    console.log('microservice-basics(js/http) tests passed')
  } finally {
    await new Promise((resolve) => server.close(resolve))
  }
})()

File: microservice-architecture/microservice-basics/go/test/test_http.go
package main

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"strconv"

	"microservice-basics-go/src"
)

func assertEqualString(expected string, actual string, message string) {
	if expected != actual {
		panic(fmt.Sprintf("%s expected=%s actual=%s", message, expected, actual))
	}
}

func main() {
	stock := 2

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		sku := r.URL.Query().Get("sku")
		qty, _ := strconv.Atoi(r.URL.Query().Get("quantity"))

		if sku == "SKU-BOOK" && qty > 0 && stock >= qty {
			stock -= qty
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte("OK"))
			return
		}

		w.WriteHeader(http.StatusConflict)
		_, _ = w.Write([]byte("NO_STOCK"))
	}))
	defer server.Close()

	client := src.NewHttpInventoryClient(server.URL)
	service := src.NewOrderService(client)

	success := service.CreateOrder("ORD-2001", "SKU-BOOK", 1)
	assertEqualString("CREATED", success.Status, "http status should be CREATED")

	failed := service.CreateOrder("ORD-2002", "SKU-BOOK", 2)
	assertEqualString("REJECTED", failed.Status, "http status should be REJECTED")

	fmt.Println("microservice-basics(go/http) tests passed")
}
